crates/db_schema/src/impls/tag.rs

```diff
@@ -13,23 +19,92 @@ use diesel::{
   ExpressionMethods,
   QueryDsl,
 };
-use diesel_async::RunQueryDsl;
-use lemmy_db_schema_file::schema::tag;
+use diesel_async::{scoped_futures::ScopedFutureExt, RunQueryDsl};
+use lemmy_db_schema_file::schema::{post_tag, tag};
 use lemmy_utils::error::{LemmyErrorExt, LemmyErrorType, LemmyResult};
+use std::collections::HashSet;
 
 impl Tag {
-  pub async fn get_by_community(
+  pub async fn read_for_community(
     pool: &mut DbPool<'_>,
-    search_community_id: CommunityId,
+    community_id: CommunityId,
   ) -> LemmyResult<Vec<Self>> {
     let conn = &mut get_conn(pool).await?;
     tag::table
-      .filter(tag::community_id.eq(search_community_id))
+      .filter(tag::community_id.eq(community_id))
       .filter(tag::deleted.eq(false))
       .load::<Self>(conn)
       .await
       .with_lemmy_type(LemmyErrorType::NotFound)
   }
+
+  pub async fn update_many(
+    pool: &mut DbPool<'_>,
+    community: &Community,
+    mut forms: Vec<TagInsertForm>,
+  ) -> LemmyResult<()> {
+    let known_tags = Tag::read_for_community(pool, community.id).await?;
+    let conn = &mut get_conn(pool).await?;
+    let new_tag_ids = forms
+      .iter()
+      .map(|tag| tag.ap_id.clone())
+      .collect::<HashSet<_>>();
+    let delete_forms = known_tags
+      .into_iter()
+      .filter(|tag| !new_tag_ids.contains(&tag.ap_id))
+      .map(|t| TagInsertForm {
+        ap_id: t.ap_id,
+        name: t.name,
+        community_id: t.community_id,
+        deleted: Some(true),
+      });
+    forms.extend(delete_forms);
+
+    conn
+      .run_transaction(|conn| {
+        async move {
+          // on conflict do update doesnt work with vec so we need a loop
+          for f in forms {
+            insert_into(tag::table)
+              .values(&f)
```

Comment:
You can definitely do insert_into with a vector of forms, you just need to use 

```rust
.set((
  column.eq(excluded.column),
  (any other columns to update here)
))
```

Reply:
Instead of doing `.set(forms)`, do the `set column.eq(excluded.column)` code I posted above. Include all the columns except the primary key.

Reply:
Okay figured it out now.